rust/template/ddlog.h

```diff
@@ -1053,35 +1059,73 @@ extern const ddlog_record* ddlog_get_struct_field(const ddlog_record* rec,
 /*
  * Create an insert command.
  *
- * `table` is the table to insert to.  `rec` is the record to insert.
- * The function takes ownership of this record.
+ * `table` - input table to insert to.
+ * `rec` - record to insert.  The function takes ownership of this record.
  *
  * Returns pointer to a new command, which can be sent to DDlog by calling
  * `ddlog_apply_updates()`.
+ *
+ * This function never fails; however the command it creates may fail to
+ * execute, causing `ddlog_apply_updates()` to return an error if:
+ * - `table` is not a valid input table id
+ * - `rec` does not match the record type of `table`
+ * - The table has a primary key and there exists a record with the same key
+ *   as `rec` in the table.
  */
 extern ddlog_cmd* ddlog_insert_cmd(table_id table, ddlog_record *rec);
 
 /*
- * Create delete-by-value command.
+ * Create an insert-or-update command that inserts a new record, deleting
+ * an existing record with the same primary key, if there is one.
+ *
+ * `table` - input table to insert to.
+ * `rec` - record to insert.  The function takes ownership of this record.
+ *
+ * Returns pointer to a new command, which can be sent to DDlog by calling
+ * `ddlog_apply_updates()`.
+ *
+ * This function never fails; however the command it creates may fail to
+ * execute, causing `ddlog_apply_updates()` to return an error if:
+ * - `table` is not a valid input table id
+ * - `table` does not have a primary key
+ * - `rec` does not match the record type of `table`
+ */
+extern ddlog_cmd* ddlog_insert_or_update_cmd(table_id table, ddlog_record *rec);
+
+/*
+ * Create a delete-by-value command.
  *
- * `table` is the table to delete from. `rec` is the record to delete.
- * The function takes ownership of this record.
+ * `table` - input table to delete from.
+ * `rec` - record to delete.  The function takes ownership of this record.
  *
  * Returns pointer to a new command, which can be sent to DDlog by calling
  * `ddlog_apply_updates()`.
+ *
+ * This function never fails; however the command it creates may fail to
+ * execute, causing `ddlog_apply_updates()` to return an error if:
+ * - `table` is not a valid input table id
+ * - `rec` does not match the record type of `table`
+ * - `table` has a primary key and record `rec` does not exist in `table`
+ *   (NOTE: for tables without a primary key, the command succeeds even if
```

Comment:
This sounds wrong

Reply:
Agreed, but at least this documents how things are working right now.